### PR TITLE
Sprint S3: Auth/MCP Cleanup & Public Surface Trims — v2.5.1 Thermo-Nuclear Patch

### PR DESCRIPTION
**Pull request by adriannoes**
## Summary

Sprint S3 of the v2.5.1 thermo-nuclear code-quality patch: behavior-preserving structural cleanup of `auth/`, `mcp/`, `transport/`, `errors`/`cli`/`a2h`, paying down debt accumulated through v2.5.0 so Adapter Lab II builds on a cleaner base. No wire-protocol, manifest schema, or public API breaking changes (deprecation shims where symbols moved).

- **auth/** collapse (Tasks 1.0+2.0+3.0): merge `claims`→`jwks`/`middleware`, `utils`→`scopes`, `lifecycle`→`identity` (16→13 files); `OAuth2Middleware` delegates to one shared `JWKSValidator` (collapses the duplicate instances deferred from S0 #237 — B4/C-2 WS-enforcement property preserved); delete `WebAuthnSelfAuthVerifier`; split `webauthn.py` into `webauthn_store.py` + `webauthn.py`; drop `agent_store_writable` mode flag, dedup `_validate_standard_claims`, drop cast, unify `_SCOPE_*`; `_validate_identity_binding` → 2-tuple.
- **mcp/** (Task 4.0): `ToolRegistration` dataclass replaces the untyped 5-tuple `_tools`; typed `JwtVerifyResult.capabilities` accessor; fold `adapters/mcp`→`asap.mcp.auth` (deprecation shim); prune dead `OpenAPIExecutionKind.ASYNC_POLLING`.
- **transport/** (Task 5.0): `transport/_state_deps.py` + migrate `*_api.py` and 9 `identity_limiter` sites to `Depends(require_identity_limiter)` (missing limiter → clean 503, not `AttributeError`); inline `codecs/`→`lambda_codec.py`; trim `transport/__all__`; drop dead `NonceStore.is_used`/`mark_used`.
- **errors/cli/a2h** (Task 6.0): collapse `RemoteFatalRPCError`/`RemoteRecoverableRPCError`→`RemoteRPCError` (`is_recoverable`; twin names kept as deprecated subclasses so chaos suites stay green with zero churn); split `cli/__init__.py` 596→91 LOC into per-command-group modules; fix `a2h.py` `RenderContent` forward-ref ordering.

## Lessons applied (from PRs #237/#239/#240 + formal v2.4→v2.5 review)

- S0 #237 deferred landmine: the two `OAuth2Middleware` instances (HTTP stack + WS `app.state`) now share ONE `JWKSValidator` — pinned by `test_oauth2_http_and_ws_share_one_jwks_validator`.
- S2 #240 HIGH finding (URN-agnostic resolve cache): `ToolRegistration` keeps capability metadata per-tool-instance; pinned by `tests/adapters/mcp/test_per_urn_isolation.py` (tool B must not inherit tool A's grant).
- S2 #240 must-fix parity: WS error-code mappings (`HandlerNotFoundError`→-32012, `ThreadPoolExhausted`→503+`retry_after_ms`) and request observability preserved through the transport migration; WS sanitize path (`str(error)` gated behind `is_debug_mode()`) not regressed by the `RemoteRPCError` collapse.
- `_send.py:533` retry decision rewritten `isinstance(rpc_exc, RemoteRecoverableRPCError)`→`rpc_exc.is_recoverable` (byte-identical retry-after).

## Commits (5, grouped by area)

1. `refactor(auth): collapse auth/ decomposition, delegate JWKS, delete WebAuthnSelfAuthVerifier (S3)`
2. `refactor(mcp): introduce ToolRegistration, fold adapters/mcp into asap.mcp.auth, prune ASYNC_POLLING (S3)`
3. `refactor(errors,cli,a2h): collapse RemoteRPCError twin, split cli, fix a2h ordering (S3)`
4. `refactor(transport): _state_deps, inline codecs, trim __init__, drop NonceStore surface (S3)`
5. `test(s3): add security-critical regression guards for v2.5.1 S3 behaviors`

## Test plan

- [x] `uv run ruff check src/asap/ tests/` — All checks passed
- [x] `uv run ruff format --check src/asap/` — 200 files formatted
- [x] `uv run mypy src/ scripts/ tests/` — no issues in 478 source files
- [x] `uv run pytest --tb=short --cov=asap --cov-fail-under=85` — **3698 passed, 12 skipped, 0 failed**, coverage **94.46%**
- [x] Public import paths verified: `asap.auth.{OAuth2Config,OAuth2Middleware}`, `asap.adapters.mcp.{protect_server,MCPAuthConfig,ProtectedMCPServer,resolve_jwt_extractor}` (shim), `asap.errors.{RemoteFatalRPCError,RemoteRecoverableRPCError,RemoteRPCError}` (deprecated subclasses), `asap.cli.{app,main,DEFAULT_SCHEMAS_DIR,export_all_schemas,_repl_namespace}` (legacy re-export), `asap.transport.lambda_codec.LAMBDA_CONTENT_TYPE`
- [x] New regression tests: shared JWKS validator, `JwtVerifyResult.capabilities` edge cases (9 parametrized), `adapters.mcp`→`mcp.auth` shim identity, per-URN capability isolation

## Notes for review

- **Documented deviations (behavior-preserving):** `auth/`=14 not 12 (13 merges + approved `webauthn_store.py` split); `cli/__init__.py`=91 LOC not ≤80 (legacy re-export fix so `tests/test_cli.py` keeps resolving); `:351` cast in `_ed25519_private_key_to_okp_key` left+flagged (input not boundary-typed); M-2 `allow_env_jwt_fallback` startup warning skipped (would corrupt subprocess `MCPClient` JSON-RPC stream).
- **No merge yet** — awaiting code review per maintainer.
- Task plan tracked in `engineering/tasks/private/v2.5.1/sprint-S3-auth-mcp-cleanup.md` (local-only, gitignored).

### CODE REVIEW

File: src/asap/transport/__init__.py [hunk 1]
```diff
@@ -23,6 +23,10 @@
     ASAPTimeoutError: Timeout error exception
     ASAPRemoteError: Remote error exception
 
+Tuning constants (``DEFAULT_*``), webhook internals, and ``start_periodic_cleanup``
```
[the remark this answers is not in the document]
Comment (adriannoes) [reply src/asap/transport/__init__.py:26]: Addressed in 065b224 — added a "Removed (import path changes)" section under `[Unreleased]` in CHANGELOG.md documenting that `start_periodic_cleanup`, the `DEFAULT_*` cache/webhook constants, and the webhook internals are no longer re-exported from `asap.transport.__init__` (still importable from `asap.transport.cache`/`asap.transport.webhook`), with the explicit downstream migration note (`from asap.transport import start_periodic_cleanup` → `from asap.transport.cache import start_periodic_cleanup`). Chose the CHANGELOG-note path over a deprecation re-export shim since the removals were grep-verified to have zero external root callers — a shim would re-advertise symbols we just trimmed. Resolving.

File: src/asap/adapters/openapi/capability_mapper.py [hunk 2]
```diff
@@ -30,11 +30,17 @@
...
 
 class OpenAPIExecutionKind(StrEnum):
-    """How upstream responses should be executed (OA-010)."""
+    """How upstream responses should be executed (OA-010).
+
+    ``async_polling`` (``202`` + ``Location``) was pruned in Sprint S3 Wave C
+    Task 4.3: it was detected and stored on :attr:`OpenAPICapability.execution_kind`
+    but never consumed by any production path (``factory`` only branches on
+    ``STREAMING``), so the variant was dead metadata. A ``202`` + ``Location``
+    operation now classifies as ``SYNC`` until a polling handler is wired.
```
[the remark this answers is not in the document]
Comment (adriannoes) [reply src/asap/adapters/openapi/capability_mapper.py:39]: Addressed in 065b224 — added a "Changed" entry under `[Unreleased]` noting the `OpenAPIExecutionKind` metadata behavior change: the dead `ASYNC_POLLING` variant is pruned and a `202 + Location` operation now classifies as `SYNC`, with the explicit caveat that external tooling reading `execution_kind` from mapped capabilities will see a different enum value. Resolving.

File: src/asap/mcp/auth/protected_server.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    def __init__(self, config: MCPAuthConfig) -> None:
+        super().__init__()
+        self._auth_config = config
+        self._jwt_extractor = resolve_jwt_extractor(config)
+        if config.hide_unauthorized_tools:
+            # MCP-MAP-004 is deferred: tools/list still returns all tools; only
+            # tools/call is gated. State this explicitly so operators do not assume
+            # unauthorized tools are hidden from discovery (formal review M-3).
+            logger.warning(
```
[the remark this answers is not in the document]
Comment (adriannoes) [reply src/asap/mcp/auth/protected_server.py:41]: Addressed in 065b224 + af4ff40. The warning now fires at `ProtectedMCPServer.__init__` construction time — but via `warnings.warn` (stderr) rather than the structlog logger (stdout). The Wave C attempt reverted a stdout log at `MCPAuthConfig` construction because the observability logger writes to `sys.stdout` (`observability/logging.py:237`), which is the same stream `run_stdio` uses for JSON-RPC, so it corrupted the subprocess `MCPClient` handshake. `protect_server()` runs inside the subprocess too, so logging at `__init__` would hit the same trap — `warnings.warn` goes to stderr by default and leaves the JSON-RPC stream intact (verified: `tests/adapters/mcp/test_stdio_integration.py` subprocess tests stay green). Pinned by `test_allow_env_jwt_fallback_emits_construction_warning` in af4ff40. Resolving.

**Comment by adriannoes** [comment]
### Code review — all threads addressed and resolved

Thanks for the review. All three "Should Fix" items and the two low-cost "Nice to Have" items are addressed in `065b224` + `af4ff40`:

| Thread | Resolution |
|--------|------------|
| **transport re-export trim** | Added a "Removed (import path changes)" CHANGELOG section under `[Unreleased]` documenting the trimmed `asap.transport.__init__` re-exports (`start_periodic_cleanup`, `DEFAULT_*`, webhook internals) with the downstream migration path. Chose the CHANGELOG-note path over a deprecation re-export shim since the removals were grep-verified to have zero external root callers. |
| **ASYNC_POLLING metadata change** | Added a "Changed" CHANGELOG entry noting `202 + Location` now classifies as `SYNC` and external tooling reading `execution_kind` will see a different enum value. |
| **doc drift** | Updated `docs/security.md` (NonceStore example → atomic `check_and_mark`) and `docs/security/self-authorization-prevention.md` (`WebAuthnSelfAuthVerifier` → `WebAuthnVerifierImpl`). |
| **M-2 `allow_env_jwt_fallback` warning** | Emitted at `ProtectedMCPServer.__init__` via `warnings.warn` (stderr) — **not** the structlog logger (stdout), because `protect_server()` runs inside the stdio subprocess and the stdout logger would corrupt the JSON-RPC stream (the same failure that forced the Wave C attempt to revert). Subprocess stdio tests stay green; pinned by `test_allow_env_jwt_fallback_emits_construction_warning`. |
| **expand S3 regression tests** | Added `test_identity_limiter_missing_returns_503_on_agent_register` (second identity route) + the M-2 warning test. |

The two larger "Nice to Have" items (`cli/__init__.py` → `cli/_compat.py` for the legacy re-exports; extracting `auth/_scope_parse.py` to remove the documented `middleware.py` inline import) are left as documented follow-ups — they're independent refactors beyond this PR's behavior-preserving scope.

**CI after fixes:** ruff ✓, mypy ✓ (200 files), pytest **3700 passed, 12 skipped**, coverage **94.45%**. No merge yet — still awaiting the final review pass.

**Comment by adriannoes** [comment]
### Nice-to-have follow-ups tracked as issues

The two larger "Nice to Have" items from the review are deferred to dedicated issues so this PR can merge behavior-preserving:

- #242 — Refactor: move legacy `asap.cli` re-exports into `cli/_compat.py` to bring `__init__.py` under the 80-LOC ceiling (low risk, ~30-40 min).
- #243 — Refactor: extract `parse_scope` into `auth/_scope_parse.py` to eliminate the documented inline import in `OAuth2Middleware` (low-medium risk, ~1-2h; includes a trade-off note — may be closed as wontfix if the team prefers to keep the documented inline-import exception).

Both are labeled `refactor`, `deferred`, `backend`.
